tdrs-backend/tdpservice/parsers/parse.py

```diff
@@ -0,0 +1,131 @@
+"""Convert raw uploaded Datafile into a parsed model, and accumulate/return any errors."""
+
+
+import os
+from . import schema_defs
+from . import validators
+
+
+def parse_datafile(datafile):
+    """Parse and validate Datafile header/trailer, then select appropriate schema and parse/validate all lines."""
+    rawfile = datafile.file
+    errors = {}
+
+    document_is_valid, document_error = validators.validate_document(rawfile)
+    if not document_is_valid:
+        errors['document'] = [document_error]
+        return errors
+
+    # get header line
+    rawfile.seek(0)
+    header_line = rawfile.readline().decode().strip()
+
+    # get trailer line
+    rawfile.seek(0)
+    rawfile.seek(-2, os.SEEK_END)
+    while rawfile.read(1) != b'\n':
+        rawfile.seek(-2, os.SEEK_CUR)
+
+    trailer_line = rawfile.readline().decode().strip('\n')
+
+    # parse header, trailer
+    header, header_is_valid, header_errors = schema_defs.header.parse_and_validate(header_line)
+    if not header_is_valid:
+        errors['header'] = header_errors
+
+    trailer, trailer_is_valid, trailer_errors = schema_defs.trailer.parse_and_validate(trailer_line)
+    if not trailer_is_valid:
+        errors['trailer'] = trailer_errors
+
+    if not header_is_valid or not trailer_is_valid:
+        return errors
```

Comment:
Per discussions with Alex, we want to proceed with parsing even if the trailer is missing/invalid. Can we amend this to only care about the header?

Reply:
just to confirm, we want to accept a document even if there's no trailer at all? that might change some things in `validate_document` (now `validate_single_header_trailer`) as well.

Reply:
That was my interpretation of what @ADPennington wants. Yes, we want to proceed anyway. cc'ing her for confirmation

Reply:
I went for the simple solution that you proposed here - don't fail preparsing validation if a trailer is missing or invalid, but leave the rest of the steps in place. multiple trailer validation still happens in `validate_single_header_trailer`, i'm open to discussing if we should remove that logic.